### PR TITLE
reticle focus feature

### LINKED ISSUE #150: reticle & focus feature
- add a small reticle at the center of the viewport 
  - white dot, black border
- reticle should indicate whether or not an object is being focused/looked at
  - reticle grows by a few px and changes color

game thread tasks:
- [x] add forward raycast, emit event to main if object hit by raycast

main thread tasks:
- [x] add reticle element to canvas
- [x] listen for reticle-focus events
- [x] modify reticle to indicate object is in focus

Unblocker for https://github.com/matrix-org/thirdroom/issues/146

### PR DESCRIPTION
**Pull request by NateTheGreatt**
resolves #150

### CODE REVIEW

File: src/plugins/reticle/reticle.game.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  const target = vec3.set(_target, 0, 0, -1);
+  vec3.transformQuat(target, target, _cameraWorldQuat);
+  vec3.scale(target, target, MAX_FOCUS_DISTANCE);
+
+  const source = mat4.getTranslation(vec3.create(), cameraMatrix);
+
+  const s: Vector3 = new Vector3().fromArray(source);
+  const t: Vector3 = new Vector3().fromArray(target);
+
+  const ray = new RAPIER.Ray(s, t);
```
Comment (robertlong) [inline src/plugins/reticle/reticle.game.ts:40]: Can we reuse this ray every frame?

File: src/plugins/reticle/reticle.game.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  const solid = true;
+  const maxToi = 4.0;
```
Comment (robertlong) [inline src/plugins/reticle/reticle.game.ts:42]: The maxToi or the final ray length should be used to determine if you can actually interact with an object based on its max interaction distance.

File: src/plugins/reticle/reticle.main.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+const reticleFocusedCss = `
+  ${reticleCss}
+  width: 8px;
+  height: 8px;
+  margin-top: -4px;
+  margin-left: -4px;
+  background: cornflowerblue;
+`;
+
+export const ReticleModule = defineModule<IMainThreadContext, ReticleModuleState>({
```
Comment (robertlong) [inline src/plugins/reticle/reticle.main.ts:35]: This should probably use a React component instead and just use the `registerMessageHandler` method inside a hook to call `setState` and then use that to pick a class.

**Review by robertlong — CHANGES_REQUESTED** [review]
I think there's a couple things that should be changed, but this is looking pretty good 👍

File: src/plugins/reticle/reticle.main.ts (hunk 3 above)
Comment (NateTheGreatt) [reply src/plugins/reticle/reticle.main.ts:35]: i considered this but do we really want to involve the vdom for such a tiny feature? i suppose yes if the reticle interfaces with the UI in a more involved way, but couldn't think of how.

Comment (robertlong) [reply src/plugins/reticle/reticle.main.ts:35]: Yeah, it's gonna be more maintainable for this, perf isn't an issue at all. The best example is the `Stats` or `Editor` components. They just get the `mainThread` reference and then do the same thing you're doing here.

Comment (robertlong) [reply src/plugins/reticle/reticle.main.ts:35]: Maybe something like this?

```typescript
export function Reticule() {
  const mainThread = useMainThreadContext();
  const [focused, setFocused] = useState<boolean>(false);

  useEffect(() => {
    const onReticleFocus = (ctx: IMainThreadContext, message: ReticleFocusMessageType) => {
      setFocused(message.focused);
    };

    return registerMessageHandler(ctx, ReticleFocusMessage, onReticleFocus);
  }, [mainThread]);

  return <div className={classNames("Reticule", { "Reticule--focused": focused })} />;
}
```

File: src/plugins/reticle/reticle.game.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  mat4.getRotation(_cameraWorldQuat, cameraMatrix);
+
+  const target = vec3.set(_target, 0, 0, -1);
+  vec3.transformQuat(target, target, _cameraWorldQuat);
+  vec3.scale(target, target, MAX_FOCUS_DISTANCE);
+
+  const source = mat4.getTranslation(vec3.create(), cameraMatrix);
+
+  const s: Vector3 = new Vector3().fromArray(source);
+  const t: Vector3 = new Vector3().fromArray(target);
```
Comment (robertlong) [inline src/plugins/reticle/reticle.game.ts:40]: There's still a few objects here being created every frame.

File: src/plugins/GrabThrowController.ts [hunk 5]
```diff
@@ -155,15 +154,14 @@ export function GrabThrowSystem(ctx: GameState) {
     const s: Vector3 = new Vector3().fromArray(source);
     const t: Vector3 = new Vector3().fromArray(target);
```
Comment (robertlong) [inline src/plugins/GrabThrowController.ts:155]: Out of scope of this PR but you may as well fix these extra allocations while you're at it

**Review by robertlong — COMMENTED** [review]
One last thing and then you're good to go!

File: src/plugins/GrabThrowController.ts (hunk 5 above)
Comment (NateTheGreatt) [reply src/plugins/GrabThrowController.ts:155]: this is only on-click but may as well

File: src/plugins/reticle/reticle.game.ts (hunk 4 above)
Comment (NateTheGreatt) [reply src/plugins/reticle/reticle.game.ts:40]: this is every frame indeed, good catch

**Review by robertlong — APPROVED** [review]
Awesome!

**Pull request merged**